src/box/xrow.c

```diff
@@ -1705,70 +1705,79 @@ xrow_decode_ballot(const struct xrow_header *row, struct ballot *ballot)
 	return -1;
 }
 
-int
-xrow_encode_register(struct xrow_header *row,
-		     const struct tt_uuid *instance_uuid,
-		     const struct vclock *vclock)
-{
-	memset(row, 0, sizeof(*row));
-	size_t size = mp_sizeof_map(2) +
-		      mp_sizeof_uint(IPROTO_INSTANCE_UUID) +
-		      mp_sizeof_str(UUID_STR_LEN) +
-		      mp_sizeof_uint(IPROTO_VCLOCK) +
-		      mp_sizeof_vclock_ignore0(vclock);
-	char *buf = (char *) region_alloc(&fiber()->gc, size);
-	if (buf == NULL) {
-		diag_set(OutOfMemory, size, "region_alloc", "buf");
-		return -1;
-	}
-	char *data = buf;
-	data = mp_encode_map(data, 2);
-	data = mp_encode_uint(data, IPROTO_INSTANCE_UUID);
-	data = xrow_encode_uuid(data, instance_uuid);
-	data = mp_encode_uint(data, IPROTO_VCLOCK);
-	data = mp_encode_vclock_ignore0(data, vclock);
-	assert(data <= buf + size);
-	row->body[0].iov_base = buf;
-	row->body[0].iov_len = (data - buf);
-	row->bodycnt = 1;
-	row->type = IPROTO_REGISTER;
-	return 0;
-}
+/**
+ * A template which can represent any replication request - join, register, ack,
+ * etc. All fields are optional - when left NULL, they are not encoded. Each
+ * specific request simply uses a subset of these fields + its own iproto
+ * request type. Meaning of each field depends on the original request type, but
+ * the iproto keys are fixed.
+ */
+struct replication_request {
+	/** IPROTO_CLUSTER_UUID. */
+	struct tt_uuid *replicaset_uuid;
+	/** IPROTO_INSTANCE_UUID. */
+	struct tt_uuid *instance_uuid;
+	/** IPROTO_VCLOCK. */
+	struct vclock *vclock;
+	/** IPROTO_ID_FILTER. */
+	uint32_t *id_filter;
+	/** IPROTO_SERVER_VERSION. */
+	uint32_t *version_id;
+	/** IPROTO_REPLICA_ANON. */
+	bool *is_anon;
+};
 
-int
-xrow_encode_subscribe(struct xrow_header *row,
-		      const struct tt_uuid *replicaset_uuid,
-		      const struct tt_uuid *instance_uuid,
-		      const struct vclock *vclock, bool anon,
-		      uint32_t id_filter)
+/** Encode a replication request template. */
+static int
+xrow_encode_replication_request(struct xrow_header *row,
+				const struct replication_request *req,
+				uint16_t type)
 {
 	memset(row, 0, sizeof(*row));
-	size_t size = XROW_BODY_LEN_MAX +
-		      mp_sizeof_vclock_ignore0(vclock);
+	size_t size = XROW_BODY_LEN_MAX;
+	if (req->vclock != NULL)
+		size += mp_sizeof_vclock_ignore0(req->vclock);
 	char *buf = (char *) region_alloc(&fiber()->gc, size);
 	if (buf == NULL) {
 		diag_set(OutOfMemory, size, "region_alloc", "buf");
 		return -1;
 	}
+	uint32_t id_filter = req->id_filter != NULL ? *req->id_filter : 0;
+	uint32_t map_size = 0;
+	map_size += req->replicaset_uuid != NULL;
+	map_size += req->instance_uuid != NULL;
+	map_size += req->vclock != NULL;
+	map_size += id_filter != 0;
+	map_size += req->is_anon != NULL;
+	map_size += req->version_id != NULL;
```

Comment:
Fixmap is suitable for maps with up to 15 entries, so I would suggest reserving one byte for a map and then coming back and filling in the map size after the element encoding is done. So that you don't have to calculate map size separately.
We could return to your current approach when and if replication_request starts holding more than 15 fields.
WDYT?

Reply:
I don't mind, changed to fixed-size map.